include/lbann/trainers/trainer.hpp

```diff
@@ -0,0 +1,157 @@
+  }
+};
+
+/** Base class for LBANN trainers. */
+class trainer {
+public:
+
+  /** Constructor. */
+  trainer(lbann_comm *comm);
+
+  /** Copy constructor. */
+  trainer(const trainer& other);
+  /** Copy assignment operator. */
+  trainer& operator=(const trainer& other);
+  /** Destructor. */
+  ~trainer();
+
+  /** Set the trainer's name; this is an arbitrary string
+   *  that may be useful in multi-trainer scenarios, e.g,
+   *  LTFB, jag
+   */
+  void set_name(std::string const& name);
+
+  /** Return the trainer's name; this is an arbitrary string
+   *  that may be useful in multi-trainer scenarios, e.g,
+   *  LTFB, jag
+   */
+  std::string get_name() const {
+    return m_name;
+  }
+
+  /** Human-readable description. */
+  description get_description() const;
+
+  /** Set up the trainer. */
+  void setup(std::unique_ptr<thread_pool> io_thread_pool);
+
+  using execution_context_key_pair_t = typename std::pair<observing_ptr<model>, execution_mode>;
+
+  execution_context_key_pair_t
+  check_and_build_execution_context(training_algorithm& alg,
+                                    observing_ptr<model> model,
+                                    execution_mode mode);
+
+  execution_context_key_pair_t
+  check_and_build_execution_context(const execution_context& c,
+                                    model& model,
+                                    execution_mode mode);
+
+  execution_context& get_execution_context(observing_ptr<model> model,
+                                                                 execution_mode mode);
+
+  execution_context& get_execution_context(execution_context_key_pair_t key);
+
+  void for_each_execution_context(std::function<bool(observing_ptr<execution_context>)>fn);
+
+  void apply(training_algorithm& alg,
+             observing_ptr<model> model,
+             execution_mode mode,
+             termination_criteria const& term_criteria);
+
+  void train(observing_ptr<model> model, El::Int num_epochs, El::Int num_batches=0);
+
+  void evaluate(observing_ptr<model> model, execution_mode mode, El::Int num_batche=0);
+
+  /** Return the I/O thread pool */
+  observing_ptr<thread_pool> get_io_thread_pool() const { return m_io_thread_pool.get(); }
+
+  /** Get the trainer's comm. */
+  inline lbann_comm *get_comm() const {
+    return m_comm;
+  }
+
+  /** Set a flag that can be used to enable / disable the background I/O activities */
+  void allow_background_io_activity(bool enable) { m_background_io_allowed = enable; }
+
+  /** Are background I/O activities enabled by the input layers */
+  bool background_io_activity_allowed() { return m_background_io_allowed; }
+
+private:
+
+  /** Give trainer a name. */
+  std::string m_name;
+
+  /** Communicator for the trainer. */
+  lbann_comm *m_comm;
+
+  /** Threads available for I/O */
+  std::unique_ptr<thread_pool> m_io_thread_pool;
+
+  /** Flag that allows input layers to fetch data in the background */
+  bool m_background_io_allowed;
```

Comment:
This field should be last. This gets into compiler optimizations and struct packing so the specifics are platform/compiler-dependent, but it's generally advisable to either list members in increasing or decreasing size (for no strong reason, I generally prefer decreasing size).